src/Roslyn.Diagnostics.Analyzers/Core/AbstractApplyTraitToClass.cs

```diff
@@ -0,0 +1,111 @@
+        {
+            var attribute = await context.TryGetRelevantNodeAsync<TAttributeSyntax>(RefactoringHelpers).ConfigureAwait(false);
+            if (attribute is null)
+            {
+                // No attribute in context
+                return;
+            }
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(context.Document);
+            if (syntaxGenerator.TryGetContainingDeclaration(attribute, DeclarationKind.Method) is null)
+            {
+                // The attribute is not applied to a method
+                return;
+            }
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var attributeType = semanticModel!.GetTypeInfo(attribute, context.CancellationToken);
+            if (attributeType.Type is not { Name: "TraitAttribute", ContainingNamespace: { Name: "Xunit", ContainingNamespace: { IsGlobalNamespace: true } } })
+                return;
+
+            var location = attribute.GetLocation();
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    RoslynDiagnosticsAnalyzersResources.ApplyTraitToContainingType,
+                    cancellationToken => ApplyTraitToClassAsync(context.Document, location.SourceSpan, cancellationToken),
+                    nameof(AbstractApplyTraitToClass<TAttributeSyntax>)));
+        }
+
+        private async Task<Document> ApplyTraitToClassAsync(Document document, TextSpan sourceSpan, CancellationToken cancellationToken)
+        {
+            var semanticModel = (await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false))!;
+
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxRoot = await syntaxTree!.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var reportedNode = syntaxRoot.FindNode(sourceSpan, getInnermostNodeForTie: true);
+            var typeDeclaration = GetTypeDeclarationForNode(reportedNode);
+            if (typeDeclaration is null)
+                return document;
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(document);
+            if (syntaxGenerator.TryGetContainingDeclaration(reportedNode, DeclarationKind.Attribute) is not { } attribute)
+            {
+                return document;
+            }
+
+            if (syntaxGenerator.TryGetContainingDeclaration(attribute, DeclarationKind.Method) is not { } method)
+            {
+                // The attribute is not applied to a method
+                return document;
+            }
+
+            var expectedAttributeData = semanticModel.GetDeclaredSymbol(method, cancellationToken).GetAttributes()
+                .Single(attributeData => attributeData.ApplicationSyntaxReference is not null && attribute.Span.Contains(attributeData.ApplicationSyntaxReference.Span));
```

Comment:
Can typing fast lead to span being changed, hence `Single(...)` can fail? I might be overthinking though :smile:

Reply:
Everything should be consistent when we get here. The `Single` check is not expected to fail, but would provide a stack trace and context in the unexpected case we failed to account for some form of attribute.